src/textual/widgets/_tabs.py

```diff
@@ -467,50 +474,54 @@ def clear(self) -> AwaitRemove:
         underline.highlight_start = 0
         underline.highlight_end = 0
         self.call_after_refresh(self.post_message, self.Cleared(self))
-        return self.query("#tabs-list > Tab").remove()
+        self.active = ""
+        return AwaitComplete(self.query("#tabs-list > Tab").remove()())
 
-    def remove_tab(self, tab_or_id: Tab | str | None) -> AwaitRemove:
+    def remove_tab(self, tab_or_id: Tab | str | None) -> AwaitComplete:
         """Remove a tab.
 
         Args:
             tab_or_id: The Tab to remove or its id.
 
         Returns:
-            An awaitable object that waits for the tab to be removed.
+            An optionally awaitable object that waits for the tab to be removed.
         """
-        if tab_or_id is None:
-            return self.app._remove_nodes([], None)
+        if not tab_or_id:
+            return AwaitComplete(self.app._remove_nodes([], None)())
+
         if isinstance(tab_or_id, Tab):
             remove_tab = tab_or_id
         else:
             try:
                 remove_tab = self.query_one(f"#tabs-list > #{tab_or_id}", Tab)
             except NoMatches:
-                return self.app._remove_nodes([], None)
-        removing_active_tab = remove_tab.has_class("-active")
+                return AwaitComplete(self.app._remove_nodes([], None)())
 
+        removing_active_tab = remove_tab.has_class("-active")
         next_tab = self._next_active
-        result_message: Tabs.Cleared | Tabs.TabActivated | None = None
-        if removing_active_tab and next_tab is not None:
-            result_message = self.TabActivated(self, next_tab)
-        elif self.tab_count == 1:
-            result_message = self.Cleared(self)
 
         remove_await = remove_tab.remove()
 
         async def do_remove() -> None:
             """Perform the remove after refresh so the underline bar gets new positions."""
             await remove_await
-            if removing_active_tab:
-                if next_tab is not None:
+
+            removal_complete = asyncio.Event()
```

Comment:
@willmcgugan Any thoughts on this - you mentioned the threat of deadlocks when using events. Does this usage look alright?

We need to wait until after the refresh before handing control back the awaiting user, since the `-active` class cannot be set until after after refresh.

Reply:
Yeah, I suspect this could deadlock.

The `call_after_refresh` will post a message to self. If you have `await self.remove_tab(...)` in a message handler then it will get stuck waiting for the refresh, which won't ever be processed.

At least I think that's what would happen. Worth testing.

I think this may suggest that we need a different implementation of `call_after_refresh`. Or maybe we can do without it in this instance.

Reply:
I'll test it out. Unfortunately I think the `call_after_refresh` is necessary here, a quick test without it caused issues. Unless there's a smarter implementation we can use here that can calculate the new tab position in advance, before the refresh happens.

Reply:
Fixed the deadlock here by waiting for the event in another coroutine, scheduling it, then including the Future in the returned AwaitComplete (via `gather`).